File: dotnet/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example1/OrderRequest.cs
namespace ChainOfResponsibilityDesignPattern.Example1;

public class OrderRequest
{
    public string OrderId { get; set; } = string.Empty;
    public string CustomerName { get; set; } = string.Empty;
    public string ProductName { get; set; } = string.Empty;
    public int Quantity { get; set; }
    public decimal TotalAmount { get; set; }
    public string? DiscountCode { get; set; }
    public bool IsApproved { get; set; }
    public List<string> ProcessMessages { get; set; } = new();

    public void AddMessage(string message)
    {
        ProcessMessages.Add(message);
    }

    public void DisplayStatus()
    {
        Console.WriteLine($"\n--- Order {OrderId} ---");
        Console.WriteLine($"Customer: {CustomerName}");
        Console.WriteLine($"Product: {ProductName} (Qty: {Quantity})");
        Console.WriteLine($"Total: ${TotalAmount:F2}");
        Console.WriteLine($"Status: {(IsApproved ? "APPROVED" : "REJECTED")}");
        Console.WriteLine("\nProcess Log:");
        foreach (var message in ProcessMessages)
        {
            Console.WriteLine($"  {message}");
        }
        Console.WriteLine();
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example1/src/OrderRequest.ts
export interface IOrderHandler {
  setNext(handler: IOrderHandler): IOrderHandler;
  handle(request: OrderRequest): void;
}

export class OrderRequest {
  orderId: string;
  customerName: string;
  productName: string;
  quantity: number;
  totalAmount: number;
  discountCode?: string;
  isApproved = false;
  readonly processMessages: string[] = [];

  constructor(init: {
    orderId: string;
    customerName: string;
    productName: string;
    quantity: number;
    totalAmount: number;
    discountCode?: string;
  }) {
    this.orderId = init.orderId;
    this.customerName = init.customerName;
    this.productName = init.productName;
    this.quantity = init.quantity;
    this.totalAmount = init.totalAmount;
    this.discountCode = init.discountCode;
  }

  addMessage(message: string): void {
    this.processMessages.push(message);
  }

  displayStatus(): void {
    console.log(`\n--- Order ${this.orderId} ---`);
    console.log(`Customer: ${this.customerName}`);
    console.log(`Product: ${this.productName} (Qty: ${this.quantity})`);
    console.log(`Total: $${this.totalAmount.toFixed(2)}`);
    console.log(`Status: ${this.isApproved ? "APPROVED" : "REJECTED"}`);
    console.log("\nProcess Log:");
    for (const msg of this.processMessages) {
      console.log(`  ${msg}`);
    }
    console.log();
  }
}
